### ISSUE TITLE
Code-guru-profiler: Unable to marshall request to JSON

### ISSUE THREAD

**Issue by KevinAtSesam** [aws/aws-sdk-java-v2#2875]
### Describe the bug

When the Code Guru wants to generate a report, it's unable to properly serialize it using Jackson.

It previously worked though, and this bug has been caued by updating a central awssdk library.

The change occured when updating the `software.amazon.awssdk::sns`  and `software.amazon.awssdk::cloudwatch` from version `2.17.72` to `2.17.92`. The Code Guru profiler is unchanged and at version `1.2.1`

### Expected behavior

I expect the serialization to work.

### Current behavior

```
ov 30, 2021 4:15:20 PM software.amazon.codeguruprofilerjavaagent.ProfilingCommand reportProfile
INFO: Attempting to report profile data: start=2021-11-30T15:10:19.732821389Z end=2021-11-30T15:15:19.681051401Z force=false memoryRefresh=false numberOfTimesSampled=300
Nov 30, 2021 4:15:20 PM software.amazon.codeguruprofilerjavaagent.flightrecorder.ObjectCountEventProcessor log
INFO: [HeapSummary] Processed 8 events.
Nov 30, 2021 4:15:21 PM software.amazon.codeguruprofilerjavaagent.ProfilingCommand submitProfilingData
INFO: Successfully reported profile
Nov 30, 2021 4:15:22 PM software.amazon.codeguruprofilerjavaagent.profile.metadata.ErrorsMetadata recordSdk
INFO: Exception while calling agent orchestration.
software.amazon.awssdk.core.exception.SdkClientException: Unable to marshall request to JSON: Can not write a field name, expecting a value
	at software.amazon.awssdk.core.exception.SdkClientException$BuilderImpl.build(SdkClientException.java:98)
	at software.amazon.awssdk.services.codeguruprofiler.transform.ConfigureAgentRequestMarshaller.marshall(ConfigureAgentRequestMarshaller.java:52)
	at software.amazon.awssdk.services.codeguruprofiler.transform.ConfigureAgentRequestMarshaller.marshall(ConfigureAgentRequestMarshaller.java:31)
	at software.amazon.awssdk.core.internal.handler.BaseClientHandler.lambda$finalizeSdkHttpFullRequest$0(BaseClientHandler.java:67)
	at software.amazon.awssdk.core.internal.util.MetricUtils.measureDuration(MetricUtils.java:51)
	at software.amazon.awssdk.core.internal.handler.BaseClientHandler.finalizeSdkHttpFullRequest(BaseClientHandler.java:66)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.doExecute(BaseSyncClientHandler.java:151)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.lambda$execute$1(BaseSyncClientHandler.java:82)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.measureApiCallSuccess(BaseSyncClientHandler.java:175)
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.execute(BaseSyncClientHandler.java:76)
	at software.amazon.awssdk.core.client.handler.SdkSyncClientHandler.execute(SdkSyncClientHandler.java:45)
	at software.amazon.awssdk.awscore.client.handler.AwsSyncClientHandler.execute(AwsSyncClientHandler.java:56)
	at software.amazon.awssdk.services.codeguruprofiler.DefaultCodeGuruProfilerClient.configureAgent(DefaultCodeGuruProfilerClient.java:324)
	at software.amazon.codeguruprofilerjavaagent.CodeGuruProfilerSDKClient.configureAgent(CodeGuruProfilerSDKClient.java:108)
	at software.amazon.codeguruprofilerjavaagent.CodeGuruProfilerSDKClient.configureAgent(CodeGuruProfilerSDKClient.java:93)
	at software.amazon.codeguruprofilerjavaagent.ProfilingCommand.callConfigureAgent(ProfilingCommand.java:183)
	at software.amazon.codeguruprofilerjavaagent.ProfilingCommand.refreshConfiguration(ProfilingCommand.java:199)
	at software.amazon.codeguruprofilerjavaagent.ProfilingCommand.startSampling(ProfilingCommand.java:143)
	at software.amazon.codeguruprofilerjavaagent.Timer.time(Timer.java:69)
	at software.amazon.codeguruprofilerjavaagent.ProfilingCommandExecutor.run(ProfilingCommandExecutor.java:78)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

### Steps to Reproduce

```
    new Profiler.Builder()
      .profilingGroupName(profile)
      .awsCredentialsProvider(StaticCredentialsProvider.create(awsCredentials))
      .withHeapSummary(true)
      .build()
      .start()
```

_Runs in the background_

### AWS Java SDK version used

2.17.92

### JDK version used

openjdk version "17.0.1" 2021-10-19 

### Operating System and version

Amazon Linux 2

**Comment by debora-ito** [comment]
Hi @KevinAtSesam thank you for reaching out.

The Java SDK team is not the owner/maintainer of the CodeGuru Profiler agent, I will contact the team behind it.
In the meantime, can you please answer some questions?
- Can you reproduce the issue reliably?
- Do you still see the error when you go back to sns and cloudwatch dependency version 2.17.72?
- Can you enable the DEBUG logs in the Profiler? See "Advanced troubleshooting" for instructions here: https://docs.aws.amazon.com/codeguru/latest/profiler-ug/troubleshooting.html

**Comment by KevinAtSesam** [comment]
Thanks for coming back to me so quickly. 

- Yes, I've been able to reproduce this both locally and in our staging environments on EC2 instances
- No, when I roll the version back, the issue disappears. This is also how I 'triaged' the issue on our staging servers.
- Here is the extra log: [codeguru-log.txt](https://github.com/aws/aws-sdk-java-v2/files/7631985/codeguru-log.txt). Sadly, I don't see much more info.

**Comment by KevinAtSesam** [comment]
Minor update; the issue is also still precent with `2.17.100`

**Comment by debora-ito** [comment]
Quick update: I'm trying to reproduce the issue, while familiarizing myself with the Code Guru Profiler Java agent.

Looking at the stacktrace, there's one call made using the AWS SDK for Java and it's `DefaultCodeGuruProfilerClient.configureAgent( )`. I'm trying to make my local test code go through that call.

**Comment by debora-ito** [comment]
Cannot reproduce using SDK version `2.17.92`. My local tests go through 
`software.amazon.awssdk.thirdparty.jackson.core.json.UTF8JsonGenerator.writeFieldName( )` successfully.

Wondering if all the SDK modules are on the same version. Can you share your dependency tree? I'm looking specially for the `third-party-jackson-core` version.

**Comment by KevinAtSesam** [comment]
Reproducing the issue with;
- AWS Sdk: 2.17.121
- awsCodeGuru: 1.2.1

I've looked into `third-party-jackson-core` to see if it's evicted somewhere. But that doesn't seem to be the case. I made a complete sbt-dependencyTree and it's massive: 650.000 lines.... Sorry in advance.

- [tree.zip](https://github.com/aws/aws-sdk-java-v2/files/7986418/tree.zip) (.tar.gz was only 220 Kb, but alas)

Going through all the references, it appears that they all use version 2.17.121, although sometimes it's truncated because the lines were getting too long. Even then, still no evictions.

I hope that this helps you in finding the cause. If there is something else you need, then don't shy away and ask.

**Comment by debora-ito** [comment]
The `software.amazon.awssdk:codeguruprofiler` version is pinned down to 2.15.4, can you try using `2.17.121`?
The version of the service modules should be the same of the core.

**Comment by KevinAtSesam** [comment]
I think that you've found the bug. When I manually define `software.amazon.awssdk:codeguruprofiler` with version `2.17.121` it works.

```
[info]     +-software.amazon.awssdk:codeguruprofiler:2.15.4 (evicted by: 2.17.121)
[info]     +-software.amazon.awssdk:codeguruprofiler:2.17.121
```

This means that there is likely an issue with the package of codeguru itself. Our SBT File after applying the fix:

```
val awsSdkVersion = "2.17.121"

[...]

// For AWS CodeGuru integration
val awsCodeGuru = "software.amazon.codeguruprofiler" % "codeguru-profiler-java-agent" % "1.2.1"
val awsCodeGuruCore = "software.amazon.awssdk" % "codeguruprofiler" % awsSdkVersion

[...]

libraryDependencies ++= Seq(
  awsCodeGuru,
  awsCodeGuruCore
),
```

Initially I didn't include `awsCodeGuruCore` since I assumed that `codeguru-profiler-java-agent` would automatically come up with the right version. This is also not mentioned in the documentation:

https://docs.aws.amazon.com/codeguru/latest/profiler-ug/enabling-the-agent-with-code.html

So, two possible actions now:
- Change the documentation to explicitly mention `software.amazon.awssdk:codeguruprofiler` in the dependencies
- Change the `software.amazon.codeguruprofiler:codeguru-profiler-java-agent` package to properly resolve the latest dependency

Is there any way how I can help further?

**Comment by debora-ito** [comment]
Glad we found the bug!

Yes, I'll reach out to the Code Guru team to add a note on the docs.

**Comment by debora-ito** [comment]
@KevinAtSesam have you tried using the Bill of Materials (BOM) with the SDK modules? You don't have to declare the version for each individual module.
https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/setup-project-gradle.html

After a quick test in my local machine I can see that if I add only add `software.amazon.codeguruprofiler:codeguru-profiler-java-agent` as a dependency when using the BOM method, the version of `software.amazon.awssdk:codeguruprofiler` is resolved to the latest one, not 2.15.4.

**Comment by KevinAtSesam** [comment]
@debora-ito mmh. I'll look into that. We use SBT to compile our application (Written in Scala) and we have it split up in multiple modules. This makes the whole dependency management a bit more involved, but it's not insurmountable.
